src/coreclr/System.Private.CoreLib/src/System/Variant.cs

```diff
@@ -78,323 +23,311 @@ internal partial struct Variant
-            _flags = flags;
-            _objref = or;
-            _data = data;
-        }
-
-        public Variant(bool val)
-        {
-            _objref = null;
-            _flags = CV_BOOLEAN;
-            _data = (val) ? bool.True : bool.False;
-        }
-
-        public Variant(sbyte val)
-        {
-            _objref = null;
-            _flags = CV_I1;
-            _data = val;
-        }
-
-        public Variant(byte val)
-        {
-            _objref = null;
-            _flags = CV_U1;
-            _data = val;
-        }
-
-        public Variant(short val)
-        {
-            _objref = null;
-            _flags = CV_I2;
-            _data = val;
-        }
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Variant_ConvertValueTypeToRecord")]
+        private static partial void ConvertValueTypeToRecord(ObjectHandleOnStack obj, out ComVariant pOle);
 
-        public Variant(ushort val)
+        internal static ComVariant GetComIPFromObjectRef(object? obj)
         {
-            _objref = null;
-            _flags = CV_U2;
-            _data = val;
+            IntPtr pUnk = GetIUnknownOrIDispatchForObject(ObjectHandleOnStack.Create(ref obj), out bool isIDispatch);
+            return ComVariant.CreateRaw(isIDispatch ? VarEnum.VT_DISPATCH : VarEnum.VT_UNKNOWN, pUnk);
         }
 
-        public Variant(char val)
-        {
-            _objref = null;
-            _flags = CV_CHAR;
-            _data = val;
-        }
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "MarshalNative_GetIUnknownOrIDispatchForObject")]
+        private static partial IntPtr GetIUnknownOrIDispatchForObject(ObjectHandleOnStack o, [MarshalAs(UnmanagedType.Bool)] out bool isIDispatch);
 
-        public Variant(int val)
+        private static object? GetObjectRefFromComIP(IntPtr pUnk)
         {
-            _objref = null;
-            _flags = CV_I4;
-            _data = val;
+            return pUnk == IntPtr.Zero ? null : Marshal.GetObjectForIUnknown(pUnk);
         }
 
-        public Variant(uint val)
+        private static unsafe object? ConvertWrappedObject(object? wrapped) => wrapped switch
         {
-            _objref = null;
-            _flags = CV_U4;
-            _data = val;
-        }
-
-        public Variant(long val)
+            null or Empty => null,
+            Enum => wrapped.GetType(),
+            IntPtr or UIntPtr => wrapped,
+            ValueType v when RuntimeHelpers.GetMethodTable(v)->IsPrimitive => null,
+            DateTime or TimeSpan or Currency => null,
+            _ => wrapped,
+        };
+
+        // Helper code for marshaling managed objects to VARIANT's
+        internal static void MarshalHelperConvertObjectToVariant(object? o, out ComVariant pOle)
         {
-            _objref = null;
-            _flags = CV_I8;
-            _data = val;
-        }
+            // Cases handled at native side: string, bool, primitives including (U)IntPtr excluding U(Int)64, array
```

Comment:
If this is the case, I would handle them in the switch and `throw` or assert so we know what is going on.

Reply:
The question is the caller from RECORD case of `MarshalHelperCastVariant`.

The pre-existing behavior is convert whatever received to VARIANT, and use `VariantChangeType` to attempt to convert it to VT_RECORD. However, `VariantChangeType` doesn't allow anything else to cast to VT_RECORD.

To determine whether an object will be casted to VT_RECORD, I'm making a call to `MarshalHelperConvertObjectToVariant` to execute all the rule in it. So these types are end up hittable and will be handled by the `IConvertible` case.

Reply:
I also have a question about observable exception throwing behavior. Getting unmarshallable types (TimeSpan, SafeHandle etc) passed into `CastVariant` will throw corresponding exception. I'm currently only passing value types because they won't create things that need clean-up in the VARIANT.

Reply:
These are great questions and I honestly have no idea at present. Testing is the only way to understand where we are.